src/main/java/slimeknights/tconstruct/plugin/jei/JEIPlugin.java

```diff
@@ -30,20 +34,58 @@
 import slimeknights.tconstruct.library.smeltery.Cast;
 import slimeknights.tconstruct.library.smeltery.CastingRecipe;
 import slimeknights.tconstruct.library.smeltery.ICastingRecipe;
+import slimeknights.tconstruct.library.tools.IToolPart;
 import slimeknights.tconstruct.shared.block.BlockTable;
 import slimeknights.tconstruct.smeltery.TinkerSmeltery;
 import slimeknights.tconstruct.smeltery.block.BlockCasting;
+import slimeknights.tconstruct.smeltery.client.GuiSmeltery;
+import slimeknights.tconstruct.smeltery.client.GuiTinkerTank;
+import slimeknights.tconstruct.smeltery.client.IGuiLiquidTank;
 import slimeknights.tconstruct.tools.TinkerTools;
 import slimeknights.tconstruct.tools.common.block.BlockToolTable;
 
 @mezz.jei.api.JEIPlugin
 public class JEIPlugin implements IModPlugin {
 
   public static IJeiHelpers jeiHelpers;
+  // crafting grid slots, integer constants from the default crafting grid implementation
+  private static final int craftOutputSlot = 0;
+  private static final int craftInputSlot1 = 1;
+  public static ICraftingGridHelper craftingGridHelper;
 
   @Override
-  public void registerItemSubtypes(ISubtypeRegistry subtypeRegistry) {
+  public void registerItemSubtypes(ISubtypeRegistry registry) {
+    TableSubtypeInterpreter tableInterpreter = new TableSubtypeInterpreter();
+    PatternSubtypeInterpreter patternInterpreter = new PatternSubtypeInterpreter();
 
+    // drying racks and item racks
+    if(TConstruct.pulseManager.isPulseLoaded(TinkerGadgets.PulseId)) {
+      registry.registerSubtypeInterpreter(Item.getItemFromBlock(TinkerGadgets.rack), tableInterpreter);
+    }
+
+    // tools
+    if(TConstruct.pulseManager.isPulseLoaded(TinkerTools.PulseId)) {
+      // tool tables
+      registry.registerSubtypeInterpreter(Item.getItemFromBlock(TinkerTools.toolTables), tableInterpreter);
+      registry.registerSubtypeInterpreter(Item.getItemFromBlock(TinkerTools.toolForge), tableInterpreter);
+
+      // tool parts
+      ToolPartSubtypeInterpreter toolPartInterpreter = new ToolPartSubtypeInterpreter();
+      for(IToolPart part : TinkerRegistry.getToolParts()) {
```

Comment:
For the future, since we're using Java8 now, we can use foreach + filter very well for things like this